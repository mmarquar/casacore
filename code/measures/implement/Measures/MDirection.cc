//# MDirection.cc:  A Measure: astronomical direction
//# Copyright (C) 1995,1996,1997
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#ifdef __GNUG__
#include <aips/Measures/Quantum.h>
typedef Quantum<Double> gpp_direction_bug1;
#endif
#include <aips/Exceptions.h>
#include <aips/Utilities/Assert.h>
#include <aips/Arrays/Vector.h>
#include <aips/RTTI/Register.h>
#include <aips/Measures/MDirection.h>

//# Constructors
MDirection::MDirection() :
  MeasBase<MVDirection,MDirection::Ref>() {}

MDirection::MDirection(const MVDirection &dt) : 
  MeasBase<MVDirection,MDirection::Ref>(dt,MDirection::DEFAULT) {}

MDirection::MDirection(const MVDirection &dt, const MDirection::Ref &rf) : 
  MeasBase<MVDirection,MDirection::Ref>(dt,rf) {}

MDirection::MDirection(const MVDirection &dt, uInt rf) : 
  MeasBase<MVDirection,MDirection::Ref>(dt,rf) {}

MDirection::MDirection(const Quantity &dt, const Quantity &dt1) : 
  MeasBase<MVDirection,MDirection::Ref>(MVDirection(dt,dt1),
					MDirection::DEFAULT) {}

MDirection::MDirection(const Quantity &dt, const Quantity &dt1,
		       const MDirection::Ref &rf) : 
  MeasBase<MVDirection,MDirection::Ref>(MVDirection(dt,dt1),rf) {}

MDirection::MDirection(const Quantity &dt, const Quantity &dt1,
		       uInt rf) : 
  MeasBase<MVDirection,MDirection::Ref>(MVDirection(dt,dt1),rf) {}

MDirection::MDirection(const Quantum<Vector<Double> > &dt) :
  MeasBase<MVDirection,MDirection::Ref>(MVDirection(dt),
					MDirection::DEFAULT) {}

MDirection::MDirection(const Quantum<Vector<Double> > &dt,
		       const MDirection::Ref &rf) : 
  MeasBase<MVDirection,MDirection::Ref>(MVDirection(dt),rf) {}

MDirection::MDirection(const Quantum<Vector<Double> > &dt,
		       uInt rf) : 
  MeasBase<MVDirection,MDirection::Ref>(MVDirection(dt),rf) {}

MDirection::MDirection(const Measure *dt) :
  MeasBase<MVDirection,MDirection::Ref>(dt) {}

MDirection::MDirection(const MeasValue *dt) :
  MeasBase<MVDirection,MDirection::Ref>(*(MVDirection*)dt,
					MDirection::DEFAULT) {}

MDirection::MDirection(const MDirection::Ref &rf) : 
  MeasBase<MVDirection,MDirection::Ref>(rf) {}

//# Destructor
MDirection::~MDirection() {}

//# Operators

//# Member functions

const String &MDirection::tellMe() const {
    return MDirection::showMe();
}

const String &MDirection::showMe() {
    static const String name("Direction");
    return name;
}

uInt MDirection::type() const {
  return Register((MDirection *)0);
}

void MDirection::assert(const Measure &in) {
  if (in.type() != Register((MDirection *)0)) {
    throw(AipsError("Illegal Measure type argument: " +
		    MDirection::showMe()));
  };
}

const String &MDirection::showType(uInt tp) {
    static const String tname[MDirection::N_Types] = {
	"J2000",
	"JMEAN",
	"JTRUE",
	"APP",
	"B1950",
	"BMEAN",
	"BTRUE",
	"GALACTIC",
	"HADEC",
	"AZEL",
        "AZELSW",
	"JNAT",
	"ECLIPTIC",
	"MECLIPTIC",
	"TECLIPTIC",
	"SUPERGAL" };
    static const String pname[MDirection::N_Planets - MDirection::MERCURY] = {
	"MERCURY",
	"VENUS",
	"MARS",
	"JUPITER",
	"SATURN",
	"URANUS",
	"NEPTUNE",
	"PLUTO",
	"SUN",
	"MOON" };

    if ((tp & MDirection::EXTRA) == 0) {
      DebugAssert(tp < MDirection::N_Types, AipsError);
      return tname[tp];
    };
    DebugAssert((tp & ~MDirection::EXTRA) < 
		(MDirection::N_Planets - MDirection::MERCURY), AipsError);
    return pname[tp & ~MDirection::EXTRA];
}

Bool MDirection::giveMe(const String &in, MDirection::Ref &mr) {
    static const Int N_name = 27;
    static const String tname[N_name] = {
	"J2000",
	"JMEAN",
	"JTRUE",
	"APP",
	"B1950",
	"BMEAN",
	"BTRUE",
	"GALACTIC",
	"HADEC",
	"AZEL",
        "AZELSW",
	"AZELNE",
	"JNAT",
	"ECLIPTIC",
	"MECLIPTIC",
	"TECLIPTIC",
	"SUPERGAL",
	"MERCURY",
	"VENUS",
	"MARS",
	"JUPITER",
	"SATURN",
	"URANUS",
	"NEPTUNE",
	"PLUTO",
	"SUN",
	"MOON" };

    static const uInt oname[N_name] = {
	MDirection::J2000,
	MDirection::JMEAN,
	MDirection::JTRUE,
	MDirection::APP,
	MDirection::B1950,
	MDirection::BMEAN,
	MDirection::BTRUE,
	MDirection::GALACTIC,
	MDirection::HADEC,
	MDirection::AZEL,
        MDirection::AZELSW,
	MDirection::AZEL,
	MDirection::JNAT,
	MDirection::ECLIPTIC,
	MDirection::MECLIPTIC,
	MDirection::TECLIPTIC,
	MDirection::SUPERGAL,
	MDirection::MERCURY,
	MDirection::VENUS,
	MDirection::MARS,
	MDirection::JUPITER,
	MDirection::SATURN,
	MDirection::URANUS,
	MDirection::NEPTUNE,
	MDirection::PLUTO,
	MDirection::SUN,
	MDirection::MOON };

    uInt i = Measure::giveMe(in, N_name, tname);

    if (i>=N_name) {
	mr = MDirection::Ref();
	return False;
    } else {
	mr = MDirection::Ref(oname[i]);
    };
    return True;
}

MDirection::GlobalTypes MDirection::globalType(uInt tp) {

    static const MDirection::GlobalTypes oname[MDirection::N_Types] = {
	MDirection::GRADEC,
	MDirection::GRADEC,
	MDirection::GRADEC,
	MDirection::GRADEC,
	MDirection::GRADEC,
	MDirection::GRADEC,
	MDirection::GRADEC,
	MDirection::GLONGLAT,
	MDirection::GHADEC,
	MDirection::GAZEL,
        MDirection::GAZEL,
	MDirection::GRADEC,
	MDirection::GLONGLAT,
	MDirection::GLONGLAT,
	MDirection::GLONGLAT,
	MDirection::GLONGLAT };
    if ((tp & MDirection::EXTRA) != 0) tp = 0;
    DebugAssert(tp < MDirection::N_Types, AipsError);

    return oname[tp];
}

Quantum<Vector<Double> > MDirection::getAngle() const {
    return (data.getAngle());
}

Quantum<Vector<Double> > MDirection::getAngle(const Unit &inunit) const {
    return (data.getAngle(inunit));
}

Measure *MDirection::clone() const {
    return (new MDirection(*this));
}
