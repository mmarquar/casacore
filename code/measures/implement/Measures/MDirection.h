//# MDirection.h: A Measure: astronomical direction
//# Copyright (C) 1995,1996,1997
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//#
//# $Id$

#if !defined(AIPS_MDIRECTION_H)
#define AIPS_MDIRECTION_H

#if defined(_AIX)
#pragma implementation ("MDirection.cc")
#endif

//# Includes
#include <aips/aips.h>
#include <aips/Measures/MeasBase.h>
#include <aips/Measures/MeasRef.h>
#include <aips/Measures/MVDirection.h>

//# Forward Declarations
class MDirection;
class MCDirection;
template <class M, class F, class MC> class MeasConvert;

//# Typedefs

// <summary>
//  A Measure: astronomical direction
// </summary>

// <use visibility=export>

// <reviewed reviewer="" date="yyyy/mm/dd" tests="tMeasure" demos="">
// </reviewed>

// <prerequisite>
//   <li> <linkto class=Measure>Measure</linkto> class 
// </prerequisite>
//
// <etymology>
// </etymology>
//
// <synopsis>
// MDirection forms a derived Measure class for a direction in space.<br>
// An MDirection can be generated from a
// <linkto class=MVDirection>MVDirection</linkto> or a pair of 
// <linkto class=Quantum>Quantities</linkto> specifying a longitudinal and a
// latitudinal angle.<br>
// The different reference types that can be used for a Direction are:
// <ul>
//  <li>MDirection::J2000 -- mean equator and equinox at J2000.0
//  <li>MDirection::JNAT --- geocentric natural frame
//  <li>MDirection::JMEAN -- mean equator and equinox at frame epoch
//  <li>MDirection::JTRUE -- true equator and equinox at frame epoch
//  <li>MDirection::APP ---- apparent geocentric position
//  <li>MDirection::B1950 -- mean epoch and ecliptic at B1950.0
//  <li>MDirection::BMEAN -- mean equator and equinox at frame epoch
//  <li>MDirection::BTRUE -- true equator and equinox at frame epoch
//  <li>MDirection::GALACTIC -- galactic coordinates
//  <li>MDirection::HADEC -- topocentric HA and declination
//  <li>MDirection::AZEL --- topocentric Azimuth and Elevation (N through E)
//  <li>MDirection::AZELSW - topocentric Azimuth and Elevation (S through W)
//  <li>MDirection::AZELNE - topocentric Azimuth and Elevation (N through E)
//  <li>MDirection::MERCURY -- the planet: has no data attached
//  <li>MDirection::VENUS
//  <li>MDirection::MARS
//  <li>MDirection::JUPITER
//  <li>MDirection::SATURN
//  <li>MDirection::URANUS
//  <li>MDirection::NEPTUNE
//  <li>MDirection::PLUTO
//  <li>MDirection::SUN
//  <li>MDirection::MOON

//  <li>MDirection::DEFAULT = J2000
// </ul>
// <p>
// Conversion between the different types is done with the standard
// <linkto class=MeasConvert>MeasConvert</linkto> class 
// (<src>MDirection::Convert</src> in this case).<br>
// For some conversion additional <linkto class=MeasFrame>MeasFrame</linkto>
// information is essential. The following list specifies which information
// is needed if the conversion goes to or from the different types:
// <ul>
//   <li><em>Epoch</em>: all but J2000, B1950, GALACTIC
//   <li><em>Positiom</em>: HADEC, AZEL
// </ul>
// Conversions are based on the IAU system of 
// <linkto class=Precession>precession</linkto> and 
// <linkto class=Nutation>nutation</linkto> (with
// IERS corrections if available); and on series expansions of the DE200
// planetary ephemeris (J system; for B sytem older expansions) for the
// <linkto class=Aberration>aberration</linkto> and the 
// <linkto class=SolarPos>solar position</linkto>.<br>
// The <em>APP</em> position has corrections for polar motion and the
// equation of equinoxes; the <em>AZEL</em> will include Earth tides and
// refraction at a later stage.<br>
// Note that conversion between B1950 and J2000 can only be approximate, and is
// based on FK4 to FK% conversion. The best conversion is to convert first
// to an apparent position at the time of observation, and convert from there
// to the other standard (the correct route will be followed).<br>
// Another problem can arise if the source has proper motion and/or radial
// velocities. These should be taken into account. An
// MCatalog class will maybe take care of that.
// <p>
// To aid in formatting of the angles without having to check all difference
// referencetypes, the following global types are provided:
// <ul>
//  <li> GRADEC for types that are probably expressed in HM,DM
//  <li> GHADEC for types that are probably expressed in +-HM,DM
//  <li> GAZEL for types that are probably expressed in +-deg,deg
//  <li> GLONGLAT for types that are probably expressed in deg,deg
// </ul>
// they can be obtained with the globalType() method. 
// </synopsis>
//
// <example>
// See <linkto module=Measures>Measures</linkto> module description for
// extensive examples.
// </example>
//
// <motivation>
// </motivation>
//
// <todo asof="1996/02/21">
//	<li>
// </todo>

class MDirection : public MeasBase<MVDirection,MeasRef<MDirection> > 
{
public:
//# Friends
// Conversion of data
    friend class MeasConvert<MDirection,MVDirection,MCDirection>;

//# Enumerations
// Types of known MDirections
// <note role=warning> The order defines the order in the translation matrix FromTo
// in the getConvert routine. Do not change the order without
// changing the array. Additions should be made before N_types, and
// an additional row and column should be coded in FromTo, and
// in showType().</note>
    enum Types {
      J2000,
      JMEAN,
      JTRUE,
      APP,
      B1950,
      BMEAN,
      BTRUE,
      GALACTIC,
      HADEC,
      AZEL,
      AZELSW,
      JNAT,
      N_Types,
      // Planets. First one should be Mercury
      MERCURY = 32,
      VENUS,
      MARS,
      JUPITER,
      SATURN,
      URANUS,
      NEPTUNE,
      PLUTO,
      SUN,
      MOON,
      N_Planets,
      // All extra bits
      EXTRA = 32,
      // Defaults
      DEFAULT=J2000,
      // Synonyms
      AZELNE=AZEL};
  // Global types
  enum GlobalTypes {
    GRADEC,
    GHADEC,
    GAZEL,
    GLONGLAT,
    N_GTypes};

//# Typedefs
// MeasRef use
    typedef class MeasRef<MDirection> Ref;
// MeasConvert use
    typedef class MeasConvert<MDirection,MVDirection,MCDirection> Convert;

//# Constructors
// <note role=tip> In the following constructors and other functions, all 
// <em>MeasRef</em> can be replaced with simple <src>Measure::TYPE</src>
// where no offsets or frames are needed in the reference. For reasons
// of compiler limitations the formal arguments had to be specified as
// <em>uInt</em> rather than the Measure enums that should be used as actual 
// arguments.</note>
// Default constructor; generates the J2000 pole direction
    MDirection();
// Create from data and reference
// <group>
    MDirection(const MVDirection &dt);
    MDirection(const MVDirection &dt, const MDirection::Ref &rf);
    MDirection(const MVDirection &dt, uInt rf);
    MDirection(const Quantity &dt, const Quantity &dt1);
    MDirection(const Quantity &dt, const Quantity &dt1, 
	       const MDirection::Ref &rf);
    MDirection(const Quantity &dt, const Quantity &dt1, 
	       uInt rf);
    MDirection(const Quantum<Vector<Double> > &dt);
    MDirection(const Quantum<Vector<Double> > &dt, 
	       const MDirection::Ref &rf);
    MDirection(const Quantum<Vector<Double> > &dt, 
	       uInt rf);
    MDirection(const Measure *dt);
    MDirection(const MeasValue *dt);
    MDirection(const MDirection::Ref &rf);
// </group>

//# Destructor
    ~MDirection();

//# Operators

//# General Member Functions
// Tell me your type ('Direction')
// <group>
    virtual const String &tellMe() const;
    static const String &showMe();
// </group>
// Tell me your reference type (as Register())
    virtual uInt type() const;
// Assert you are a direction
    static void assert(const Measure &in);
// Tell me the global type (like GRADEC) for tp (tp like MDirection::J2000)
  static MDirection::GlobalTypes globalType(uInt tp);
// Translate reference code tp (should be given as e.g. MDirection::J2000)
    static const String &showType(uInt tp);
// Translate string to reference code
    Bool giveMe(const String &in, MDirection::Ref &mr);

// Get Measure data
// <group>
    Quantum<Vector<Double> > getAngle() const;
    Quantum<Vector<Double> > getAngle(const Unit &inunit) const;
// </group>

// Make a copy
// <group>
    virtual Measure *clone() const;
// </group>

private:

//# Data

//# Member functions

};

#endif

