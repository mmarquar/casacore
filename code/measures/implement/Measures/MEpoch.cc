//# MEpoch.cc: A Measure: instant in time
//# Copyright (C) 1995,1996,1997
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#ifdef __GNUG__
#include <aips/Measures/Quantum.h>
typedef Quantum<Double> gpp_mepoch_bug1;
#endif
#include <aips/Exceptions/Error.h>
#include <aips/Utilities/Assert.h>
#include <aips/RTTI/Register.h>
#include <aips/Measures/MEpoch.h>

//# Constructors
MEpoch::MEpoch() : 
  MeasBase<MVEpoch,MEpoch::Ref>() {}

MEpoch::MEpoch(const MVEpoch &dt) : 
  MeasBase<MVEpoch,MEpoch::Ref>(dt,MEpoch::DEFAULT) {}

MEpoch::MEpoch(const MVEpoch &dt, const MEpoch::Ref &rf) : 
  MeasBase<MVEpoch,MEpoch::Ref>(dt,rf) {}

MEpoch::MEpoch(const MVEpoch &dt, uInt rf) : 
  MeasBase<MVEpoch,MEpoch::Ref>(dt,rf) {}

MEpoch::MEpoch(const Quantity &dt) : 
  MeasBase<MVEpoch,MEpoch::Ref>(dt,MEpoch::DEFAULT) {}

MEpoch::MEpoch(const Quantity &dt, const MEpoch::Ref &rf) : 
  MeasBase<MVEpoch,MEpoch::Ref>(dt,rf) {}

MEpoch::MEpoch(const Quantity &dt, uInt rf) : 
  MeasBase<MVEpoch,MEpoch::Ref>(dt,rf) {}

MEpoch::MEpoch(const Measure *dt) :
  MeasBase<MVEpoch,MEpoch::Ref>(dt) {}

MEpoch::MEpoch(const MeasValue *dt) :
  MeasBase<MVEpoch,MEpoch::Ref>(*(MVEpoch*)dt,
				MEpoch::DEFAULT) {}

//# Destructor
MEpoch::~MEpoch() {}

//# Operators

//# Member functions

const String &MEpoch::tellMe() const {
  return MEpoch::showMe();
}

const String &MEpoch::showMe() {
  static const String name("Epoch");
  return name;
}

uInt MEpoch::type() const {
  return Register((MEpoch *)0);
}

void MEpoch::assert(const Measure &in) {
  if (in.type() != Register((MEpoch *)0)) {
    throw(AipsError("Illegal Measure type argument: " +
		    MEpoch::showMe()));
  };
}

const String &MEpoch::showType(uInt tp) {
  static const String tname[MEpoch::N_Types] = {
    "LAST",
    "LMST",
    "GMST1",
    "GAST",
    "UT1",
    "UT2",
    "UTC",
    "TAI",
    "TDT",
    "TCG",
    "TDB",
    "TCB"};
  DebugAssert((tp & ~MEpoch::EXTRA) < MEpoch::N_Types, AipsError);
  return tname[tp & ~MEpoch::EXTRA];
}

Bool MEpoch::giveMe(const String &in, MEpoch::Ref &mr) {
  static const Int N_name = 17;
  static const String tname[N_name] = {
    "LAST",
    "LMST",
    "GMST1",
    "GAST",
    "UT1",
    "UT2",
    "UTC",
    "TAI",
    "TDT",
    "TCG",
    "TDB",
    "TCB",
    "IAT",
    "GMST",
    "TT",
    "ET",
    "UT"};
  
  static const uInt oname[N_name] = {
    MEpoch::LAST,
    MEpoch::LMST,
    MEpoch::GMST1,
    MEpoch::GAST,
    MEpoch::UT1,
    MEpoch::UT2,
    MEpoch::UTC,
    MEpoch::TAI,
    MEpoch::TDT,
    MEpoch::TCG,
    MEpoch::TDB,
    MEpoch::TCB,
    MEpoch::TAI,
    MEpoch::GMST1,
    MEpoch::TDT,
    MEpoch::TDT,
    MEpoch::UT1};
  
  uInt i = Measure::giveMe(in, N_name, tname);
  if (i>=N_name) {
    mr = MEpoch::Ref();
    return False;
  } else {
    mr = MEpoch::Ref(oname[i]);
  };
  return True;
}

Quantity MEpoch::get(const Unit &inunit) const {
  return (data.getTime().get(inunit));
}

Measure *MEpoch::clone() const {
  return (new MEpoch(*this));
}
