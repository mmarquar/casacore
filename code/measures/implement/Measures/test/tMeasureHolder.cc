//# tMeasureHolder.cc: This program tests MeasureHolder
//# Copyright (C) 1998
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes

#include <aips/aips.h>
#include <aips/Exceptions.h>
#include <aips/Measures/Quantum.h>
#include <aips/Measures/MDirection.h>
#include <aips/Measures/MDoppler.h>
#include <aips/Measures/MEpoch.h>
#include <aips/Measures/MFrequency.h>
#include <aips/Measures/MPosition.h>
#include <aips/Measures/MRadialVelocity.h>
#include <aips/Measures/QuantumHolder.h>
#include <aips/Measures/MeasureHolder.h>
#include <aips/Arrays/Vector.h>
#include <aips/Containers/Record.h>
#include <aips/Glish/GlishRecord.h>

Int main() {

  try {

    cout << "----------------------------------------------------" << endl;
    cout << "Test MeasureHolder  " << endl;
    cout << "----------------------------------------------------" << endl;

    String error;
    MeasureHolder q00;
    MDirection x00(Quantity(30, "deg"), Quantity(-40, "deg"));
    Record y00;
    GlishRecord z00;

    cout << "Input value:              " << x00 << " (" <<
	  x00.getRefString() << ")" << endl;
    if (MeasureHolder(x00).toRecord(error, y00)) {
      if (q00.fromRecord(error, y00)) {
	cout <<"Record output value:      " << q00.asMeasure() << " (" <<
	  q00.asMeasure().getRefString() << ")" << endl;
      } else {
	cout << "From error: " << error << endl;
      };
    } else {
      cout << "To error: " << error << endl;
    };
    if (MeasureHolder(x00).toRecord(error, z00)) {
      if (q00.fromRecord(error, z00)) {
	cout <<"Glish output value:       " << q00.asMeasure() << " (" <<
	  q00.asMeasure().getRefString() << ")" << endl;
      } else {
	cout << "From error: " << error << endl;
      };
    } else {
      cout << "To error: " << error << endl;
    };

    MeasureHolder q01 = q00;
    MeasureHolder q02(q00);
    if (q00.asMDirection().getValue().getValue()(0) !=
	q01.asMDirection().getValue().getValue()(0) ||
	q00.asMDirection().getValue().getValue()(0) !=
        q02.asMDirection().getValue().getValue()(0)) {
      cout << "Error in copy constructor or assignment" << endl;
    };

    cout << "Is measure:        " << q00.isMeasure() << endl;
    cout << "Is direction:      " << q00.isMDirection() << endl;
    cout << "Is empty:          " << q00.isEmpty() << endl;
    cout << "Is epoch:          " << q00.isMEpoch() << endl;
    cout << "Is doppler:        " << q00.isMDoppler() << endl;
    cout << "Is position:       " << q00.isMPosition() << endl;
    cout << "Is frequency:      " << q00.isMFrequency() << endl;
    cout << "Is radialvelocity: " << q00.isMRadialVelocity() << endl;
    cout << "As measure:        " << q00.asMeasure() << endl;
    cout << "As direction:      " << q00.asMDirection() << endl;

    cout << "Error expected:" << endl;
    cout << "Input value:              " << x00 << " (" <<
	  x00.getRefString() << ")" << endl;
    if (MeasureHolder(x00).toRecord(error, y00)) {
      y00.renameField("units", RecordFieldId("refer"));
      if (q00.fromRecord(error, y00)) {
	cout <<"Record output value:      " << q00.asMeasure() << " (" <<
	  q00.asMeasure().getRefString() << ")" << endl;
      } else {
	cout << "From error: " << error << endl;
      };
    } else {
      cout << "To error: " << error << endl;
    };

  } catch (AipsError x) {
    cout << x.getMesg() << endl;
  } end_try;

  try {

    String error;
    MeasureHolder q00;
    MDirection x00(Quantity(30, "deg"), Quantity(-40, "deg"));
    Record y00;
    GlishRecord z00;

    cout << "Input value:              " << x00 << " (" <<
	  x00.getRefString() << ")" << endl;
    if (MeasureHolder(x00).toRecord(error, y00)) {
      if (q00.fromRecord(error, y00)) {
	cout <<"Record output value:      " << q00.asMeasure() << " (" <<
	  q00.asMeasure().getRefString() << ")" << endl;
      } else {
	cout << "From error: " << error << endl;
      };
    } else {
      cout << "To error: " << error << endl;
    };
    cout << "As epoch:      " << q00.asMEpoch() << endl;

  } catch (AipsError x) {
    cout << x.getMesg() << endl;
  } end_try;

  try {
    {
      String error;
      MeasureHolder q00;
      MEpoch x00(Quantity(30456, "d"));
      Record y00;
      GlishRecord z00;
      
      cout << "Input value:              " << x00 << " (" <<
	x00.getRefString() << ")" << endl;
      if (MeasureHolder(x00).toRecord(error, y00)) {
	if (q00.fromRecord(error, y00)) {
	  cout <<"Record output value:      " << q00.asMeasure() << " (" <<
	    q00.asMeasure().getRefString() << ")" << endl;
	} else {
	  cout << "From error: " << error << endl;
	};
      } else {
	cout << "To error: " << error << endl;
      };
      cout << "As epoch:      " << q00.asMEpoch() << endl;
    }      
    {
      String error;
      MeasureHolder q00;
      MDoppler x00(Quantity(30456, "m/s"));
      Record y00;
      GlishRecord z00;
      
      cout << "Input value:              " << x00 << " (" <<
	x00.getRefString() << ")" << endl;
      if (MeasureHolder(x00).toRecord(error, y00)) {
	if (q00.fromRecord(error, y00)) {
	  cout <<"Record output value:      " << q00.asMeasure() << " (" <<
	    q00.asMeasure().getRefString() << ")" << endl;
	} else {
	  cout << "From error: " << error << endl;
	};
      } else {
	cout << "To error: " << error << endl;
      };
      cout << "As Doppler:      " << q00.asMDoppler() << endl;
    }      
    {
      String error;
      MeasureHolder q00;
      MFrequency x00(Quantity(30456, "MHz"));
      Record y00;
      GlishRecord z00;
      
      cout << "Input value:              " << x00 << " (" <<
	x00.getRefString() << ")" << endl;
      if (MeasureHolder(x00).toRecord(error, y00)) {
	if (q00.fromRecord(error, y00)) {
	  cout <<"Record output value:      " << q00.asMeasure() << " (" <<
	    q00.asMeasure().getRefString() << ")" << endl;
	} else {
	  cout << "From error: " << error << endl;
	};
      } else {
	cout << "To error: " << error << endl;
      };
      cout << "As Frequency:      " << q00.asMFrequency() << endl;
    }      
    {
      String error;
      MeasureHolder q00;
      MPosition x00(Quantity(6, "Mm"), Quantity(20, "deg"),
		    Quantity(30, "deg"));
      Record y00;
      GlishRecord z00;
      
      cout << "Input value:              " << x00 << " (" <<
	x00.getRefString() << ")" << endl;
      if (MeasureHolder(x00).toRecord(error, y00)) {
	if (q00.fromRecord(error, y00)) {
	  cout <<"Record output value:      " << q00.asMeasure() << " (" <<
	    q00.asMeasure().getRefString() << ")" << endl;
	} else {
	  cout << "From error: " << error << endl;
	};
      } else {
	cout << "To error: " << error << endl;
      };
      cout << "As Position:      " << q00.asMPosition() << endl;
    }      
    {
      String error;
      MeasureHolder q00;
      MRadialVelocity x00(Quantity(30456, "m/s"));
      Record y00;
      GlishRecord z00;
      
      cout << "Input value:              " << x00 << " (" <<
	x00.getRefString() << ")" << endl;
      if (MeasureHolder(x00).toRecord(error, y00)) {
	if (q00.fromRecord(error, y00)) {
	  cout <<"Record output value:      " << q00.asMeasure() << " (" <<
	    q00.asMeasure().getRefString() << ")" << endl;
	} else {
	  cout << "From error: " << error << endl;
	};
      } else {
	cout << "To error: " << error << endl;
      };
      cout << "As RadialVelocity:      " << q00.asMRadialVelocity() << endl;
    }      

  } catch (AipsError x) {
    cout << x.getMesg() << endl;
  } end_try;

  exit(0);
}
