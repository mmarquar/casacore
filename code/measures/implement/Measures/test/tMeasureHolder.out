----------------------------------------------------
Test MeasureHolder  
----------------------------------------------------
Input value:              Direction: [0.663414, 0.383022, -0.642788] (J2000)
Record output value:      Direction: [0.663414, 0.383022, -0.642788] (J2000)
Glish output value:       Direction: [0.663414, 0.383022, -0.642788] (J2000)
Is measure:        1
Is direction:      1
Is empty:          0
Is epoch:          0
Is doppler:        0
Is position:       0
Is frequency:      0
Is radialvelocity: 0
As measure:        Direction: [0.663414, 0.383022, -0.642788]
As direction:      Direction: [0.663414, 0.383022, -0.642788]
Error expected:
Input value:              Direction: [0.663414, 0.383022, -0.642788] (J2000)
From error: Illegal Measure record in MeasureHolder::fromRecord

Input value:              Direction: [0.663414, 0.383022, -0.642788] (J2000)
Record output value:      Direction: [0.663414, 0.383022, -0.642788] (J2000)
As epoch:      Empty or wrong MeasureHolder for asMEpoch
Input value:              Epoch: 30456::00:00:00.0000 (UTC)
Record output value:      Epoch: 30456::00:00:00.0000 (UTC)
As epoch:      Epoch: 30456::00:00:00.0000
Input value:              Doppler: 0.00010159 (RADIO)
Record output value:      Doppler: 0.00010159 (RADIO)
As Doppler:      Doppler: 0.00010159
Input value:              Frequency: 3.0456e+10 (LSRK)
Record output value:      Frequency: 3.0456e+10 (LSRK)
As Frequency:      Frequency: 3.0456e+10
Input value:              Position: [4.88279e+06, 1.77719e+06, 3e+06] (ITRF)
Record output value:      Position: [4.88279e+06, 1.77719e+06, 3e+06] (ITRF)
As Position:      Position: [4.88279e+06, 1.77719e+06, 3e+06]
Input value:              Radialvelocity: 30456 (LSRK)
Record output value:      Radialvelocity: 30456 (LSRK)
As RadialVelocity:      Radialvelocity: 30456
